heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java

```diff
@@ -39,12 +39,62 @@
 import com.twitter.heron.spi.utils.ShellUtils;
 
 public final class SchedulerUtils {
-  public static final int PORTS_REQUIRED_FOR_EXECUTOR = 9;
   public static final int PORTS_REQUIRED_FOR_SCHEDULER = 1;
   public static final String SCHEDULER_COMMAND_LINE_PROPERTIES_OVERRIDE_OPTION = "P";
 
   private static final Logger LOG = Logger.getLogger(SchedulerUtils.class.getName());
 
+  /**
+   * Enum that defines the type of ports that an heron executor needs
+   */
+
+  public enum ExecutorPort {
+    MASTER_PORT("master", true),
+    TMASTER_CONTROLLER_PORT("tmaster-ctl", true),
+    TMASTER_STATS_PORT("tmaster-stats", true),
+    SHELL_PORT("shell-port", true),
+    METRICS_MANAGER_PORT("metrics-mgr", true),
+    SCHEDULER_PORT("scheduler", true),
+    METRICS_CACHE_MASTER_PORT("metrics-cache-m", true),
+    METRICS_CACHE_STATS_PORT("metrics-cache-s", true),
+    CHECKPOINT_MANAGER_PORT("ckptmgr", true);
+
+    private final String name;
+    private final boolean required;
+
+    ExecutorPort(String name, boolean required) {
+      this.name = name;
+      this.required = required;
+    }
+
+    public String getName() {
+      return name;
+    }
+
+    public boolean isRequired() {
+      return required;
+    }
+
+    public static String getPort(ExecutorPort executorPort,
+                                  Map<ExecutorPort, String> portMap) {
+      if (!portMap.containsKey(executorPort) && executorPort.isRequired()) {
+        throw new RuntimeException("Required port " + executorPort.getName() + " not provided");
+      }
+
+      return portMap.get(executorPort);
+    }
+
+    public static int getRequiredNumOfPorts() {
```

Comment:
In the current approach each scheduler seeks for a specific port and adds its value. The scheduler will not fail If a new `required` port is added and the scheduler is not updated. I have a suggestion. Lets add a method to return all required ports, `Set<ExecutorPort> getRequiredPorts`. All schedulers would iterate on this set to provide the required values. If a new port is added, the scheduler code will fail with an error.

Reply:
In the current implementation, a exception will be thrown if the scheduler doesn't set all of the required ports.  The method "executorCommandArgs" calls the "getPort" method in "ExecutorPort" to get all the ports.  If the port is required and missing a runtime exception will be thrown:

https://github.com/twitter/heron/pull/2516/files#diff-2f8f616b79cedacafb019678d260770fR81

Reply:
If a new port required port is added here and retrieved in the "executorCommandArgs" method then an exception will be thrown.